File: 003-length-of-longest-substring/solution.go
package main

import "fmt"

func lengthOfLongestSubstring(s string) int {
	l, r := 0, 0
	maxLen := 0
	counting := make([]int, 256)
	
	for r < len(s) {
		counting[s[r]]++
		for counting[s[r]] > 1 {
			counting[s[l]] -= 1
			l++
		}
		maxLen = max(maxLen, r-l+1)
		r++
	}
	return maxLen
}
func main() {
	// s := "ddvdf"
	s2 := "pzwwkaew"
	// fmt.Println("Ans = ", lengthOfLongestSubstring(s2))
	fmt.Println("Ans = ", lengthOfLongestSubstring2(s2))
}

/*
代碼執行:
go run ./003-length-of-longest-substring

我們可以確定r指針會一直往前走, 遇到重複字元時, 要開始移動l
但是l的移動軌跡要怎麼設計才是這題的困難點

最直覺的做法是 l = r, l--  但這些都不是答案, 而是要用for一步一步把l 移到正確位置上, 如上面代碼所示

自己寫的時候也是卡在這裡, 寫不出來
*/

File: 003-length-of-longest-substring/solution.js
/**
 * LeetCode Problem 003 - Longest Substring Without Repeating Characters
 * https://leetcode.com/problems/longest-substring-without-repeating-characters/
 * 
 * 題目：給定一個字符串，找出不含重複字符的最長子串的長度
 */

/**
 * @param {string} s
 * @return {number}
 */
function lengthOfLongestSubstring(s) {
    const charMap = new Map();
    let maxLength = 0;
    let left = 0;

    for (let right = 0; right < s.length; right++) {
        const char = s[right];

        if (charMap.has(char)) {
            left = Math.max(left, charMap.get(char) + 1);
        }

        charMap.set(char, right);
        maxLength = Math.max(maxLength, right - left + 1);
    }

    return maxLength;
}

// 測試用例
if (import.meta.url === `file://${process.argv[1]}`) {
    const testCases = [
        { input: 'abcabcbb', expected: 3 },      // "abc"
        { input: 'bbbbb', expected: 1 },         // "b"
        { input: 'pwwkew', expected: 3 },        // "wke"
        { input: 'au', expected: 2 },            // "au"
        { input: '', expected: 0 }               // empty string
    ];

    console.log('Running tests for Problem 003...\n');
    testCases.forEach((testCase, index) => {
        const result = lengthOfLongestSubstring(testCase.input);
        const passed = result === testCase.expected;
        const status = passed ? '✓' : '✗';
        console.log(`Test ${index + 1} ${status}`);
        console.log(`  Input: "${testCase.input}"`);
        console.log(`  Expected: ${testCase.expected}, Got: ${result}\n`);
    });
}

export default lengthOfLongestSubstring;
